source/darwin/dispatch_queue_event_loop.c

```diff
@@ -0,0 +1,686 @@
+    context->io_dispatch_loop = dispatch_loop;
+    dispatch_loop->context = context;
+
+    loop->vtable = &s_vtable;
+
+    return loop;
+
+clean_up:
+    if (dispatch_loop) {
+        if (dispatch_loop->dispatch_queue) {
+            dispatch_release(dispatch_loop->dispatch_queue);
+        }
+        s_dispatch_event_loop_destroy(loop);
+    } else {
+        aws_mem_release(alloc, loop);
+    }
+    return NULL;
+}
+
+static void s_dispatch_queue_destroy_task(void *context) {
+    struct dispatch_loop *dispatch_loop = context;
+    s_rlock_dispatch_loop_context(dispatch_loop->context);
+
+    s_lock_cross_thread_data(dispatch_loop);
+    dispatch_loop->synced_data.suspended = true;
+    dispatch_loop->synced_data.current_thread_id = aws_thread_current_thread_id();
+    dispatch_loop->synced_data.is_executing = true;
+
+    // swap the cross-thread tasks into task-local data
+    struct aws_linked_list local_cross_thread_tasks;
+    aws_linked_list_init(&local_cross_thread_tasks);
+    aws_linked_list_swap_contents(&dispatch_loop->synced_data.cross_thread_tasks, &local_cross_thread_tasks);
+    s_unlock_cross_thread_data(dispatch_loop);
+
+    aws_task_scheduler_clean_up(&dispatch_loop->scheduler); /* Tasks in scheduler get cancelled*/
+    while (!aws_linked_list_empty(&local_cross_thread_tasks)) {
+        struct aws_linked_list_node *node = aws_linked_list_pop_front(&local_cross_thread_tasks);
+        struct aws_task *task = AWS_CONTAINER_OF(node, struct aws_task, node);
+        task->fn(task, task->arg, AWS_TASK_STATUS_CANCELED);
+    }
+
+    s_lock_cross_thread_data(dispatch_loop);
+    dispatch_loop->synced_data.is_executing = false;
+    s_unlock_cross_thread_data(dispatch_loop);
+
+    s_runlock_dispatch_loop_context(dispatch_loop->context);
+    s_dispatch_event_loop_destroy(dispatch_loop->base_loop);
+}
+
+static void s_destroy(struct aws_event_loop *event_loop) {
+    AWS_LOGF_TRACE(AWS_LS_IO_EVENT_LOOP, "id=%p: Destroying Dispatch Queue Event Loop", (void *)event_loop);
+    struct dispatch_loop *dispatch_loop = event_loop->impl_data;
+
+    /* make sure the loop is running so we can schedule a last task. */
+    s_run(event_loop);
+
+    /* cancel outstanding tasks */
+    dispatch_async_and_wait_f(dispatch_loop->dispatch_queue, dispatch_loop, s_dispatch_queue_destroy_task);
+
+    AWS_LOGF_TRACE(AWS_LS_IO_EVENT_LOOP, "id=%p: Releasing Dispatch Queue.", (void *)event_loop);
+}
+
+static int s_wait_for_stop_completion(struct aws_event_loop *event_loop) {
+    (void)event_loop;
+
+    return AWS_OP_SUCCESS;
+}
+
+static void s_try_schedule_new_iteration(struct dispatch_loop_context *loop, uint64_t timestamp);
+
+static int s_run(struct aws_event_loop *event_loop) {
+    struct dispatch_loop *dispatch_loop = event_loop->impl_data;
+
+    s_lock_cross_thread_data(dispatch_loop);
+    if (dispatch_loop->synced_data.suspended) {
+        AWS_LOGF_INFO(AWS_LS_IO_EVENT_LOOP, "id=%p: Starting event-loop thread.", (void *)event_loop);
```

Comment:
Since dispatch loops are created in a running state when we retrieve it from Apple's framework, we should add this log to the point where the dispatch loop is started in the `aws_event_loop_new_with_dispatch_queue` function. This one will never get called on a newly created and run dispatch loop but we want the log that one was created and Started for debugging purposes.